### PR TITLE
[Blockstore] Fix the computation of throttling delay for compaction and cleanup

### PR DESCRIPTION
**Pull request by gy2411**
There was a bug in the logic of computation of throttling delay for cleanup and compaction. Real delay duration was slightly different from the duration that is supposed to be by design. In particular, if we set  maxExecTimePerSecond=1000ms, the delay was nonzero; this is strange.

In this pr we:
- Fix the bug: replace `(TDuration::Seconds(1) - execTime) * throttlingFactor` by `TDuration::Seconds(1) * throttlingFactor - execTimeForLastSecond`
- Deduplicate code of throttle for compaction and cleanup
- Add unit tests

### Issue
#5815

### CODE REVIEW

File: cloud/blockstore/libs/storage/partition/model/background_ops_throttling.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        return {};
+    }
+
+    auto delay = minDelay;
+    if (maxExecTimePerSecond) {
+        const auto throttlingFactor =
+            static_cast<double>(execTimeForLastSecond.GetValue()) /
+            maxExecTimePerSecond.GetValue();
+        const auto throttleDelay =
+            TDuration::Seconds(1) * throttlingFactor - execTimeForLastSecond;
```
[the remark this answers is not in the document]
Comment (gy2411) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.cpp]: Disagree: `execTimeForLastSecond` has type `TDuration`, and `TDuration` can not be negative: https://github.com/ydb-platform/nbs/blob/main/util/datetime/base.h#L220.

If execTimeForLastSecond is zero, there is no delay. If maxExecTimePerSecond is >= 1 second, there is also no delay. Both these corner cases are checked in unit tests.

File: cloud/blockstore/libs/storage/partition/model/background_ops_throttling_ut.cpp [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        const auto minDelay = TDuration::MilliSeconds(10);
+        const auto delay = CalculateBackgroundOpThrottleDelay(
+            TDuration::Zero(),
+            TDuration::MilliSeconds(100),
+            minDelay,
+            TDuration::Seconds(1));
+
+        UNIT_ASSERT_VALUES_EQUAL(minDelay, delay);
+    }
+
```
Comment (SvartMetal) [inline cloud/blockstore/libs/storage/partition/model/background_ops_throttling_ut.cpp:70]: Let's add a test where execTimeForLastSecond is 1 sec or greater

File: cloud/blockstore/libs/storage/partition/model/background_ops_throttling_ut.cpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                execTime,
+                maxExec,
+                TDuration::Zero(),
+                TDuration::Hours(1));
+
+            UNIT_ASSERT_VALUES_EQUAL(TDuration::MilliSeconds(225), delay);
+            CheckBudgetInvariant(execTime, delay, maxExec);
+        }
+    }
+
```
Comment (SvartMetal) [inline cloud/blockstore/libs/storage/partition/model/background_ops_throttling_ut.cpp:152]: Excessive \n

**Review by SvartMetal — CHANGES_REQUESTED** [review]

File: cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+// Background ops (compaction, cleanup) are throttled by limiting how much
+// tablet-thread time they are allowed to consume per second
+// (maxExecTimePerSecond). The delay is derived from execution time of the
+// previous run (execTimeForLastSecond).
+//
+// For example, assume that maxExecTimePerSecond is 50ms. If the operation spent
+// 25ms in the tablet thread, the next operation is postponed by
+// 475ms (500ms − 25ms). If it spent 100ms, the next one is postponed by
+// 1.9s (2s − 100ms), and so on.
```
Comment (nisqatsi) [inline cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: maxExecTimePerSecond = 500 ms
execTimeForLastSecond = 1000 ms
throttlingFactor = 1000 / 500 = 2
throttleDelay = 1000*2 - 1000 = 1000 ms - получается, что может начать работать НАМНОГО раньше, чем в описанном варианте с maxExecTimePerSecond = 50 ms и execTimeForLastSecond = 100 ms с тем же самым throttlingFactor = 2.
Как-то странно получается. Очевидно, что при maxExecTimePerSecond = 500 для полной выработки лимита потребуется больше времени, чем при maxExecTimePerSecond = 50, но откладывать исполнение во втором варианте на интервал в 2 раза больше кажется неправильным.
Или я какой-то особой логики не понимаю?

Comment (nisqatsi) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: По идее в обоих вариантах нужно отложить на ~2 секунды, т.к. уже выработал лимит за следующую секунду.

"- execTimeForLastSecond" в формуле расчета throttleDelay фактически снижает postpone тем больше, чем больше выработал за предыдущую, в т.ч. с превышением лимита - положительная обратная связь вместо отрицательной.

Если execTimeForLastSecond означает "сколько отработали за текущую, еще не завершившуюся секунду":
* при execTimeForLastSecond < maxExecTimePerSecond идет использование времени в пределах разрешенного, следующее выполнение нужно до конца интервала от начала текущей секунды, пропорционального отработанной части лимита. **Но при этом нужно знать, в какой момент текущей секунды мы производим вычисление**
* при execTimeForLastSecond >= maxExecTimePerSecond: "количество целых занятых секунд минус одна" + "время до конца текущей секунды" + "время, занятое неполным остатком от секунды, в которую будет разрешено выполнение", рассчитанное по предыдущему варианту для execTimeForLastSecond_1 = execTimeForLastSecond % maxExecTimePerSecond

Если execTimeForLastSecond означает "сколько отработали за последнюю секунду" и вычисляется (сдвигается) гранулярно (окнами по частям секунды):
* фактически вычисление всегда производим "в конце текущей секунды"
* при execTimeForLastSecond < maxExecTimePerSecond время текущей секунды недоиспользовано, тротлить вроде как еще не нужно (вернется minDelay, по которому делается clamp)
* при execTimeForLastSecond >= maxExecTimePerSecond был перерасход за текущую секунду - нужно отложить на пропорциональную перерасходу часть секунды: throttleDelay = 1000 * ( execTimeForLastSecond / maxExecTimePerSecond - 1) - для перерасхода в 2 раза будет отложено на 1 секунду, т.е. в среднем выйдет ровно максимальный расход за текущую и следующую секунды

Comment (gy2411) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: Возможно, сбивает с толку имя execTimeForLastSecond -- оно кажется неудачным. Думаю, стоит переименовать эту переменную. **Это именно время исполнение одной последней операции (а не всех операций в течение секунды)** -- об этом написано тут: https://github.com/ydb-platform/nbs/blob/main/cloud/blockstore/libs/storage/partition/part_state.h#L625. И аналогично для cleanup: https://github.com/ydb-platform/nbs/blob/main/cloud/blockstore/libs/storage/partition/part_state.h#L964.

Логика такая: maxExecTimePerSecond определяет долю времени, которе таблетка может заниматься операцией. При троттлинге мы смотрим, сколько заняло последнее исполнение операции, и делаем такой delay, чтобы на отрезке из операции+задержки соотношение было ровно таким, какое нужно.

В случае maxExecTimePerSecond = 500 ms и execTimeForLastSecond = 1000 ms:
- Таблетка может компактить 50% времени
- Операция исполнялась 1 секунду => далем задержку в 1 секунду. Тогда получается, что на интервале операции+задержки таблетка компактила как раз 50% времени.

В случае maxExecTimePerSecond = 50 ms и execTimeForLastSecond = 100 ms:
- Таблетка может компактить 5% времени
- Операция исполнялась 100 миллисекнд => надо отсупить ещё на 1900 миллисекунд, так как 100ms / 2000 ms = 5%.

Такие циклы операция+задержка повторяются раз за разом, и получается, что совокупно таблетка занимается компакшеном не более, чем нужную долю времени.

Comment (gy2411) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: > "- execTimeForLastSecond" в формуле расчета throttleDelay фактически снижает postpone тем больше, чем больше выработал за предыдущую, в т.ч. с превышением лимита - положительная обратная связь вместо отрицательной.

Если подставить в формулу определение throttlingFactor, получаем
`execTimeForLastSecond * (1 sec / maxExecTimePerSecond - 1)`. То есть обратная связь всегда неотрицательная. Она положительная при maxExecTimePerSecond < 1 sec. Иначе она нулевая (как как тип TDuration всегда неотрицательный).

Comment (gy2411) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: В общем, на мой взгляд, с логикой здесь всё в порядке. Но увидеть её в этом коде не очень просто. Подумаю, может, можно как-то более ясно написать этот код.

Comment (nisqatsi) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: > Подумаю, может, можно как-то более ясно написать этот код.

Да, и переменную тогда понятнее назвать - lastOperationExecTime или что-то подобное.

И лучше явно описать эту логику, которую ты написал выше, в комментарии над функцией CalculateBackgroundOpThrottleDelay(). Тогда и проверить реализацию на соответствие заявленной логике можно будет, и в будущем у тех, кто будет дальше развивать/поддерживать, вопросов не возникнет "а что тут задумывалось и как это работает / должно работать", и не придется восстанавливать логику из кода.

File: cloud/blockstore/libs/storage/partition/model/background_ops_throttling_ut.cpp (hunk 2 above)
Comment (gy2411) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling_ut.cpp:70]: Added these test cases

File: cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h (hunk 4 above)
Comment (gy2411) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: Renamed the variable `execTimeForLastSecond` into `lastOperationExecTime` and improved the comment.

I will also try make some more improvements: rewrite the code of the CalculateBackgroundOpThrottleDelay function more clearly, resolve this TODO and rename the `GetCleanupExecTimeForLastSecond` method https://github.com/ydb-platform/nbs/blob/main/cloud/blockstore/libs/storage/partition/part_state.h#L963.

But I think it will be better to do this in the separate pr.

Comment (nisqatsi) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: Да, все ок.

```
permittedExecutionPart = maxExecTimePerSecond / 1000
throttlingFactor = lastOperationExecTime / maxExecTimePerSecond
throttleDelay = 1000 * throttlingFactor - lastOperationExecTime
throttledExecutionPart = lastOperationExecTime / (throttleDelay + lastOperationExecTime)

invariant: permittedExecutionPart == throttledExecutionPart

maxExecTimePerSecond / 1000 == lastOperationExecTime / (throttleDelay + lastOperationExecTime)
maxExecTimePerSecond / 1000 == lastOperationExecTime / (1000 * throttlingFactor - lastOperationExecTime + lastOperationExecTime)
maxExecTimePerSecond / 1000 == lastOperationExecTime / (1000 * throttlingFactor)
maxExecTimePerSecond / 1000 == lastOperationExecTime / (1000 * lastOperationExecTime / maxExecTimePerSecond)
maxExecTimePerSecond / 1000 == (lastOperationExecTime * maxExecTimePerSecond) / (1000 * lastOperationExecTime )
maxExecTimePerSecond / 1000 == (maxExecTimePerSecond) / (1000)
```

Добавь плиз формулу для `throttledExecutionPart` (то, что у тебя в `CheckBudgetInvariant()` вычисляется и проверяется) в коммент - по ней точно понятно, чего мы ожидаем.

Comment (nisqatsi) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: Вот более простое и понятное вычисление:

```
permittedExecutionPart = maxExecTimePerSecond / 1000
permittedExecutionInterval = lastOperationExecTime / permittedExecutionPart
throttleDelay = permittedExecutionInterval - lastOperationExecTime

где permittedExecutionInterval - интервал, на котором lastOperationExecTime
составляет ровно permittedExecutionPart от размера этого интервала
```

Доказательство:

```
invariant: permittedExecutionPart == throttledExecutionPart

throttledExecutionPart = lastOperationExecTime / (throttleDelay + lastOperationExecTime)

permittedExecutionPart == lastOperationExecTime / (throttleDelay + lastOperationExecTime)
permittedExecutionPart == lastOperationExecTime / (permittedExecutionWindow - lastOperationExecTime + lastOperationExecTime)
permittedExecutionPart == lastOperationExecTime / permittedExecutionWindow
permittedExecutionPart == lastOperationExecTime / (lastOperationExecTime / permittedExecutionPart)
permittedExecutionPart == lastOperationExecTime * permittedExecutionPart / (lastOperationExecTime)
permittedExecutionPart == permittedExecutionPart
```

Comment (gy2411) [reply cloud/blockstore/libs/storage/partition/model/background_ops_throttling.h]: Спасибо за ценные замечания и советы! Поправил код и комменарий.

**Review by nisqatsi — APPROVED** [review]

**Review by SvartMetal — APPROVED** [review]
